### ISSUE TITLE
Verify that `PointerRepository` does not have to be thread-safe

### ISSUE THREAD

**Issue by elfenpiff** [eclipse-iceoryx/iceoryx#1701]
## Brief feature description

In a this discussion it was unraveled that the PointerRepository is not-thread-safe: https://github.com/eclipse-iceoryx/iceoryx/pull/1699#discussion_r984338244

@elBoberido drafted the following possible scenario:

* thread A is started
* thread B is started
* thread A registers a pointer
* thread B registers a pointer -> is this a valid use case
* thread B uses a relative pointer -> do we need to sync the memory

Do we require memory synchronizations here or even place some kind of thread-safe mechanism? This has to be investigated further.

In my opinion the drafted scenario is a valid use case but I think that iceoryx does not use the pointer repository in that fashion. A startup looks like:
* runtime registers at roudi
* roudi sends a list with all shared memory segments to the application
* one thread in the application receive this list and registers all segments in the pointer repository
* since the runtime init is the first call threads will start to pop up after the registration is finalized

Furthermore, the pointer repository is an internal construct. So I think that we currently do not have a bug but since the expected behavior differs from the actual one we require a solution for the scenario drafted by @elBoberido . Which includes
* threadsafe pointer registration
* sync between threads

Or we try to avoid this by defining a strict contract. The problem is `unregister`, what happens when a thread unregisters a segment while others are still using them (or the relative pointers on them).

**Comment by elfenpiff** [comment]
cc @MatthiasKillat

**Comment by MatthiasKillat** [comment]
@elfenpiff @elBoberido I have to verify some details, but in the start up phase (shared memory mapping) registration happens from the same thread and causes writes to the repository. Afterwards the repository does not change and is only read from.

So as long as the data written is synchronized it works. However, this is not strictly enforced I believe (it just wors by chance I assume due to cache updates taking place regardless).

To what degree it has to be thread-safe is not a trivial question and depends on how it is used. Adding a mutex is very bad for readers, as resolving any relative pointer goes through this repository (and would require a lock then).

I believe some synchronization with atomic combined with some approach as in the `PolymorphicHandler` (no mutex, still synced via atomics unless also broken) would be the right thing.

This would still be fast and now not rely on synchronization by chance at startup. I would have to look into the details to really optimize this. 

It is also important whether unregister is really supposed to be called at runtime (currently it is not, only in tests and during shutdown). Unregister is essentially a change (write) to the repo, hence it would again require synchronization.

Maybe the `StaticLifetimeGuard` can be part of the solution to synchronize the instance (it is a singleton). However, it cannot guarantee that the instance itself is thread-safe to the degree stated above.

**Comment by elfenpiff** [comment]
@MatthiasKillat 

> I would have to look into the details to really optimize this.

I would postpone the task until we know that we still require the `PointerRepository` in the refactored version of iceoryx. It maybe possible that we can handle the communication without it.

**Comment by gpalmer-latai** [comment]
So correct me if I'm wrong but, since the `RelativePointer` class is now a part of the public API, anyone could in theory use it, which means that anyone can call `registerPtr` from anywhere, which means that there are theoretically valid scenarios where a race condition might occur. 

Realistically, I have a hard time coming up with a reason why someone would map shared memory in different threads concurrently and call this, but in theory if that happened the interleaving could result in e.g. one shared memory region mapping being overwritten by the other.

**Comment by elBoberido** [comment]
You are correct, this would be possible. I think that's why I raised my concerns last year but forgot about it in the meantime. The way we use it should be safe but currently nobody prevents one to just call `registerPtr`. I think we need to make this more obvious and maybe do not have it so prominently exposed. An option to prevent calling it from anywhere anytime would be to introduce something like a `finalize` function after which a call to `registerPtr` would fail.

cc @MatthiasKillat

**Comment by gpalmer-latai** [comment]
I wonder if you can use an atomic for the segment id such that when registering a new segment you do a `fetch_add` and it is guaranteed to be unique. The other functions that take an id as an argument could just check that they are less than the current atomic value.

**Comment by gpalmer-latai** [comment]
Ah but that does not work when registering with specific id's out of order... unless you also protect the individual entries with an `atomic<bool>`. And actually that might be all you need.

**Comment by elBoberido** [comment]
I think it would be possible to make it thread safe but that would also involve some costs we would have to pay even if there aren't multiple threads. We just have to decide which scenario we want to support and how to prevent synchronization issues on the scenarios we don't support.

**Comment by elBoberido** [comment]
Oh, I think I confused something. Have to think more about it when it's not so late :sweat_smile:
